sdks/python/apache_beam/runners/interactive/interactive_environment.py

```diff
@@ -130,6 +139,17 @@ def __init__(self, cache_manager=None):
           'You have limited Interactive Beam features since your '
           'ipython kernel is not connected any notebook frontend.')
 
+  @property
+  def options(self):
+    """A reference to the global interactive options.
+
+    Provided to avoid import loop or excessive dynamic import. All internal
+    Interactive Beam modules should access interactive_beam.options through
+    this property.
+    """
+    from apache_beam.runners.interactive.interactive_beam import options
+    return options
```

Comment:
Why is the options defined globally in a different file, but setter is here?

Reply:
The `options` instantiated in `interactive_beam` is to expose the `getters` and `setters` of configurable fields such as `enable_capture_replay`, `capture_duration` and `capturable_sources` to the `interactive beam` user.
The user has all necessary `getters`, `setters` and docstrings by looking at the `interactive_beam` module without the complexity of their underlying implementation details (such as `__repr__`, and how options and their utilities are grouped together).

Then inside all internal `interactive beam` modules, to access the fields configured by the user, since we don't want any module to depend on `interactive_beam` module to avoid import loops, we can only do dynamic importing. It's going to be messy if we just dynamic import `interactive_beam` (the module that is supposed to be used by end user) everywhere. So this property (no `setter` given) does the dynamic import once in this single place and all internal modules will depend on `interactive_environment` module to access whatever configuration the user might have set.

Reply:
If all the getters/setters are available in `interactive_beam`, why do the user need to use this setter here?

Reply:
This is a getter, not a setter, and it's used by internal modules, not users.
Users always do:
```
# Static import once.
from apache_beam.runners.interactive import interactive_beam as ib
# Use it everywhere.
ib.options.<setter/getter>
```

The purpose is to simplify below usages:
In an `interactive beam` internal module, when accessing options:
```
# Dynamic import when needed. Because no internal module should depend on 
# `interactive_beam` in a static way to avoid an import loop. Also the `interactive_beam` module
# provides APIs to the end user, should not be mixed into the implementation.
# In a local scope:
from apache_beam.runners.interactive.interactive_beam import options
use(options.getter)
...
# In another local scope:
from apache_beam.runners.interactive.interactive_beam import options
use(options.xxx)
...
```
Since most modules accessing global states have `interactive_environment` as a static dependency, providing this getter in `interactive_environment` simplified the usage to:
```
# Static import once, most modules have this as dependency already.
from apache_beam.runners.interactive import interactive_environment as ie
# In a local scope:
use(ie.current_env().options.getter)
...
# In another local scope:
use(ie.current_env().options.getter)
```